### DISCUSSION TITLE
How to query or save the raw json/text data without custom type casting when using query builder?

### DISCUSSION THREAD

**Discussion by suerta-git** [mikro-orm/mikro-orm#2769] [Q&A] [answered]
Hi there,

I have a custom type which's column type in DB is json now.

And to test if the custom type works well, I'd like to prepare some raw text data in DB, then restore it using orm to see if the raw data is casted to JS value correctly.

So for the preparing raw data step, I expect it should be a pure manual process and won't involve any custom type logic. 
I am using query builder to simplify my work. But I found if I write codes like this:
```js
const qb = repository.createQueryBuilder()
qb.update({
    customTypeProperty: rawData,
  })
  .where({ id: xxx })
```
It seems to involve the type casting method (JSValue -> DatabaseValue) during this query, which is what I don't want to see.

So is there a way to save the raw data directly using query builder? Thank you.

---

By the way, when I'd like to query the raw data from DB using query builder like this:

```js
const result = await repository
  .createQueryBuilder()
  .select('customTypeProperty')
  .where({ id: xxx })
  .execute('get', false)
```

the result I got is a deserialized js object instead of raw json string. I am not sure if it's just a default behavior something like auto-deserializing? Or it also has a similar problem as above: also involved custom type casting logic during querying data?

**Comment by B4nan** [comment, marked as answer]
`qb.unsetFlag(QueryFlag.CONVERT_CUSTOM_TYPES)`

> the result I got is a deserialized js object instead of raw json string. I am not sure if it's just a default behavior something like auto-deserializing? Or it also has a similar problem as above: also involved custom type casting logic during querying data?

That is driver specific, e.g. `pg` will do it automatically, sqlite not. Nothing that we would do at the ORM level (better say we handle this in the mapped types for drivers that don't do it automatically, so by unsetting this flag, you should get strings with sqlite, but not with other drivers).

For postgres I believe its possible to override this behaviour via https://github.com/brianc/node-pg-types/ but it might break some internals, I would not suggest doing this as it affects _all_ properties globally.

**Reply by suerta-git** [reply]
Thanks for your reply. 

Please correct me if I misunderstand, you mean the data flow is about 
```
DB (raw data) --> Driver (parsing data) --> Custom type (converting to custom type)
```

If I unset the flag, the converting in custom type is disabled. So I can get the result from the driver.

But for different drivers there are different behaviors. For json data in DB, `sqlite` returns strings but `pg` returns deserialized js objects.

**Reply by B4nan** [reply]
Yes, and by driver I mean the 3rd party dependency, not MikroORM code. So literally `sqlite3` or `pg` packages.

Namely JSON also gets parsed in hydration, but that happens only when hydrating object embeddables, not with normal properties.

**Reply by suerta-git** [reply]
By the way, if the `pg` driver performs auto-serializing and auto-deserializing for `json` data, does that mean I can just return serializable js objects as `DBType` instead of strings when implementing a custom type?

```ts
class CustomType extends Type<JSType, APlainObject> {}
```

**Reply by B4nan** [reply]
For JSON custom types I would suggest extending the ORM `JsonType`, that way it should do what's needed for all drivers.
